File: Wikipedia/Code/WMFRotationRespectingNavigationController.m
#import "WMFRotationRespectingNavigationController.h"

@implementation WMFRotationRespectingNavigationController

- (BOOL)shouldAutorotate {
    if (self.presentedViewController && ![self.presentedViewController isKindOfClass:[UIAlertController class]]) {
        return self.presentedViewController.shouldAutorotate;
    } else if (self.topViewController) {
        return self.topViewController.shouldAutorotate;
    } else {
        return YES;
    }
}

- (UIInterfaceOrientationMask)supportedInterfaceOrientations {
    if (self.presentedViewController && ![self.presentedViewController isKindOfClass:[UIAlertController class]]) {
        return self.presentedViewController.supportedInterfaceOrientations;
    } else if (self.topViewController) {
        return self.topViewController.supportedInterfaceOrientations;
    } else {
        return UIInterfaceOrientationMaskAll;
    }
}

- (UIInterfaceOrientation)preferredInterfaceOrientationForPresentation {
    if (self.presentedViewController && ![self.presentedViewController isKindOfClass:[UIAlertController class]]) {
        return self.presentedViewController.preferredInterfaceOrientationForPresentation;
    } else if (self.topViewController) {
        return self.topViewController.preferredInterfaceOrientationForPresentation;
    } else {
        return UIInterfaceOrientationPortrait;
    }
}

@end

File: Wikipedia/Code/WMFRotationRespectingNavigationController.swift

import UIKit

public class WMFRotationRespectingNavigationController: UINavigationController {
    
    public override func shouldAutorotate() -> Bool {
        if let vc = self.topViewController {
            return vc.shouldAutorotate()
        }else{
            return super.shouldAutorotate()
        }
    }
    public override func supportedInterfaceOrientations() -> UIInterfaceOrientationMask {
        if let vc = self.topViewController {
            return vc.supportedInterfaceOrientations()
        }else{
            return super.supportedInterfaceOrientations()
        }
    }
    
    public override func preferredInterfaceOrientationForPresentation() -> UIInterfaceOrientation {
        if let vc = self.topViewController {
            return vc.preferredInterfaceOrientationForPresentation()
        }else{
            return super.preferredInterfaceOrientationForPresentation()
        }
    }
}
